src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.cs

```diff
@@ -53,37 +55,49 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                 Range = ProtocolConversions.TextSpanToRange(span, text)
             };
 
-            var commands = await lspClient.RequestAsync(LSP.Methods.TextDocumentCodeAction, codeActionParams, cancellationToken).ConfigureAwait(false);
+            var commands = await lspClient.RequestAsync(LSP.Methods.TextDocumentCodeAction.ToLSRequest(), codeActionParams, cancellationToken).ConfigureAwait(false);
             if (commands == null)
             {
                 return;
             }
 
             foreach (var command in commands)
             {
-                if (command is LSP.Command lspCommand)
+                if (TryParseJson(command, out LSP.Command lspCommand))
                 {
                     // The command can either wrap a Command or a CodeAction.
                     // If a Command, leave it unchanged; we want to dispatch it to the host to execute.
                     // If a CodeAction, unwrap the CodeAction so the guest can run it locally.
                     var commandArguments = lspCommand.Arguments.Single();
 
-                    // Unfortunately, older liveshare hosts use liveshare custom code actions instead of the LSP code action.
-                    // So determine which one to pass on.
-                    if (commandArguments is LSP.CodeAction lspCodeAction)
+                    if (TryParseJson(commandArguments, out LSP.CodeAction lspCodeAction))
                     {
                         context.RegisterRefactoring(new RoslynRemoteCodeAction(document, lspCodeAction.Command, lspCodeAction.Edit, lspCodeAction.Title, lspClient));
                     }
-                    else if (commandArguments is LiveShareCodeAction liveshareCodeAction)
-                    {
-                        context.RegisterRefactoring(new RoslynRemoteCodeAction(document, liveshareCodeAction.Command, liveshareCodeAction.Edit, liveshareCodeAction.Title, lspClient));
-                    }
                     else
                     {
                         context.RegisterRefactoring(new RoslynRemoteCodeAction(document, lspCommand, lspCommand?.Title, lspClient));
                     }
                 }
             }
         }
+
+        private static bool TryParseJson<T>(object json, out T t)
+        {
+            t = default;
+            if (json == null)
+            {
+                return true;
+            }
+            try
+            {
+                t = ((JObject)json).ToObject<T>();
+                return true;
+            }
+            catch (JsonException)
```

Comment:
why catch?

Reply:
It's possible that the JObject cannot be serialized into the desired .NET type. This happens because we don't own the other half of this code. The "new" client (this code) could be connecting to an "old" server (the code we don't control). The old server may return a type that isn't what the new client expects.

Reply:
1. There's no 'try' form of this so that we can avoid noisy exceptions (which suck for debugging)?
2. If we get a bad message, is it appropriate to just silently swallow?  What's the user experience here?  Does something just silently fail?  Will they know what's up?

Reply:
1. Correct. AFAICT, this is the recommended way to do this in newtonsoft json. When debugging, this exception should never get thrown; it would only happen with mismatched versions of Live Share talking to each other.
2. Yes, I believe it's appropriate to silently fail here. The user experience is that they don't see any code actions. Since we make this request essentially on every click, any other experience would be pretty noisy. And as mentioned, this is just to cover the corner case of mismatched Live Share clients talking to each other.

Reply:
> Yes, I believe it's appropriate to silently fail here. The user experience is that they don't see any code actions

That's a really bad user experience.  the user will interpret this as the product being broken.  It will also not be able to provide any information to them (or us) as to why this is happening.

We have mechanisms to do better here.  For example:

1. we can file a watson when this happens so we have real data on if this is happening.
2. we can show the 'goldbar' to the user to let them know about this.
3. we can disable a specific fix if it crashes.

In fact, i think all of the above will happen if you just don't catch these exceptions.  If there's a problem, why not use the existing infrastructure in roslyn to handle this?  That seems much better than an effective 'catch all' which can degrade the user experience.

Reply:
I agree with the concerns. However, I think it's out-of-scope; this isn't new behavior introduced by my change. My change should behave identically to the previous behavior. I'm just replacing the strict type checking with a json deserialization "type check".

Reply:
Added https://github.com/dotnet/roslyn/projects/45#card-24375494 for me to followup here.